data/remote/src/main/java/com/ku_stacks/ku_ring/remote/sse/SSEDataSource.kt

```diff
@@ -0,0 +1,18 @@
+package com.ku_stacks.ku_ring.remote.sse
+
+import kotlinx.coroutines.flow.Flow
+
+interface SSEDataSource {
+    /**
+     * 쿠링봇 SSE 세션을 열고, [Flow] 포맷으로 반환한다.
+     *
+     * @param query 질문 내용
+     * @param token 사용자 FCM 토큰
+     *
+     * @return 세션을 정상적으로 연 경우 non-null [Flow], 오류가 발생한 경우 null
+     */
+    suspend fun openKuringBotSSESession(
+        query: String,
+        token: String,
+    ): Flow<String>?
+}
```

Comment:
```suggestion
interface SSEDataSource {
    /**
     * 쿠링봇 SSE 세션을 열고, [Flow] 포맷으로 반환한다.
     *
     * @param query 질문 내용
     * @param token 사용자 FCM 토큰
     *
     * @return 세션을 정상적으로 연 경우 non-null [Flow], 오류가 발생한 경우 null
     */
    fun openKuringBotSSESession(
        query: String,
        token: String,
    ): Flow<String>?
}
```

이 점 참고해주세요!

<img width="751" alt="스크린샷 2024-07-24 오후 10 01 49" src="https://github.com/user-attachments/assets/d6595035-f13b-412f-9ae5-4a3032b78fec">

Reply:
## 원인 설명

먼저 `suspend`가 붙은 이유를 말씀드리면, ktor의 sse 함수가 모두 suspend입니다. 단순히 세션을 열고 반환하는 `sseSession()`과, 세션에서 `block`을 실행하는 `sse()` 모두 suspend입니다.

![image](https://github.com/user-attachments/assets/53301bcf-3140-45df-a04a-57dd787adb1c)

제가 의도했던 로직은
1. `SSEDataStore`에서 SSE 스트림을 `Flow<String>`으로 반환
2. `SSERepository`에서 빈 문자열을 공백으로 치환하는 `Flow<String>`을 반환
3. `KuringBotViewModel`에서 `Flow<String>`을 받아 응답/처리 구현

이었습니다. [예시 코드](https://github.com/ku-ring/ktor-sse-test/tree/main/src/main/kotlin)도 작성해 두었습니다.

## 해결법?

일단 보시는 것처럼 suspend를 뗄 수는 없고, 대신 `sse()` 함수를 이용하여 로직을 람다식으로 받아야 할 것 같습니다.

Reply:
그런데.. 서버에서 질문 횟수를 초과했을 때 json 응답을 보낼 수도 있다고 합니다. 즉 SSE와 json을 동시에 받을 수 있어야 하는데요

이건 다음 PR에서 대응하겠습니다

Reply:
https://github.com/ku-ring/KU-Ring-Android/pull/277/commits/5332bec5fd84ca893c58a17c8b8675923c660f58 에서 `Flow<String>`을 반환하는 대신 콜백을 매개변수로 받는 방식을 구현했습니다.

Reply:
https://github.com/ku-ring/KU-Ring-Android/pull/277/commits/f78279f653a5150ae6ae334825b84188ac8074ab 에서 content type에 따라 SSE와 json을 모두 처리하는 로직을 구현했습니다.

원래는 다음 PR에서 처리할 예정이었으나, 세부 구현은 여기서 마무리하고 다음 PR에서는 repository 로직에만 집중하려고 합니다.

Reply:
https://github.com/ku-ring/KU-Ring-Android/pull/277/commits/4b424005d14faac6ea2cbc3be07fd15c570b6b61 에서 `Flow`를 반환하는 대신 콜백을 매개변수로 받도록 수정했습니다.

_cf. 쿠링봇 쿼리에 대한 응답은 항상 SSE 스트림으로 내려오는 것으로 결정되었습니다. Json 등 다른 포맷은 사용하지 않습니다._

Reply:
순서대로 쿼리 내용이 범위를 벗어난 경우, 정상 응답, 요청 횟수를 초과했을 때의 응답입니다. 전부 수신 양호.

![image](https://github.com/user-attachments/assets/5d3efe62-d91c-4109-b07c-b83910621c83)